refactor(d.ts): Export type and refactor code

kamontat opened the pull request.

jonschlinkert commented:
Nice work! We're reviewing the code. I'm still learning typescript, so feedback from others would be great as well. Thanks!

g-plane commented:
@kamontat Good job!

g-plane commented on index.d.ts:
```diff
@@ -1,148 +1,345 @@
+declare namespace EnquirerStatic {
+  namespace Caller {
+    type NoParam<R> = () => R;
+    type OneParam<T, R = T> = (value: T) => R;
```
Why the default type of the type parameter `R` is `T`?

kamontat replied:
The `R` is for the return but I set as an optional since somethings the method will return as the same as input

g-plane commented on index.d.ts:
```diff
@@ -1,148 +1,345 @@
-      | (PromptOptions | ((this: Enquirer) => PromptOptions))[]
-  ): Promise<T>;
+declare namespace EnquirerStatic {
+  namespace Caller {
```
Maybe we can merge `Caller` with `UnionType` namespace as a new namespace called `TypeUtils`?

kamontat replied:
yes.

g-plane commented on index.d.ts:
```diff
@@ -1,148 +1,345 @@
+        NumberOptions<N> {}
+  }
+
+  type PromptOptions<N extends string> =
```
Enquirer allows custom prompt, not only built-in prompts.

kamontat replied:
I already add support to custom prompt, but the prompt must extend `BaseTypeOptions`

kamontat commented:
I will try to solve it problem, but I have another work at least until end of the year. so...

kamontat commented:
after end of this year, can you notify me again, otherwise I may forget about this project.

g-plane commented:
@kamontat Are you still active on this?

kamontat commented:
Yes, sorry for late update.

Cyberuben commented:
What is the status of this PR? I'm looking for TS definitions and it seems pointless writing my own if these definitions work

kamontat commented:
This is working, but not 100%

jpike88 commented:
Can this be made a priority? TS defintion support is the standard nowadays, wouldn't be a bad idea either to convert this project to TS

g-plane commented:
@kamontat What's the progress now? Are you still active on this?

cdaringe commented:
Hey friends, what is needed to progress this forward? If there's something concrete that needs to happen, can you enumerate it? Getting some of these types corrected would be a great help

kamontat commented:
@cdaringe There are a lot changes since last time I working on this PR (both enquirer and Typescript).
I can continue on this if @enquirer team still needs this. I didn't use/follow enquirer anymore so I not sure what is the plan for Typescript definitions.

There are a lot of Typescript definitions open on this repository.
